pkg/core/native/notary.go

```diff
@@ -0,0 +1,345 @@
+	if deposit == nil {
+		if amount.Cmp(big.NewInt(2*transaction.NotaryServiceFeeFactor)) < 0 {
+			return stackitem.NewBool(false)
+		}
+		deposit = &state.Deposit{Amount: new(big.Int)}
+	} else {
+		if till < deposit.Till {
+			return stackitem.NewBool(false)
+		}
+	}
+	err := n.GAS.TransferInternal(ic, from, n.Hash, amount)
+	if err != nil {
+		return stackitem.NewBool(false)
+	}
+	// TODO: check that `till` > current chain's height?
+	deposit.Amount.Add(deposit.Amount, amount)
+	deposit.Till = till
+	if err := n.putDepositFor(ic.DAO, deposit, to); err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", to.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+*/
+
+// withdraw sends all deposited GAS for "from" address to "to" address.
+func (n *Notary) withdraw(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	to := toUint160(args[1])
+	if to.Equals(util.Uint160{}) {
+		to = from
+	}
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if ic.Chain.BlockHeight() < deposit.Till {
+		return stackitem.NewBool(false)
+	}
+	err := n.GAS.TransferInternal(ic, from, to, deposit.Amount)
+	if err != nil {
+		return stackitem.NewBool(false)
+	}
+	if err := ic.DAO.DeleteStorageItem(n.ContractID, append([]byte{prefixDeposit}, from.BytesBE()...)); err != nil {
+		panic(fmt.Errorf("failed to remove withdrawn deposit for %s from the storage: %w", from.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+
+// balanceOf returns deposited GAS amount for specified address.
+func (n *Notary) balanceOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		return stackitem.NewBigInteger(big.NewInt(0))
+	}
+	return stackitem.NewBigInteger(deposit.Amount)
+}
+
+// expirationOf Returns deposit lock height for specified address.
+func (n *Notary) expirationOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		panic(fmt.Errorf("deposit for %s not found", acc.StringBE()))
+	}
+	return stackitem.NewBigInteger(big.NewInt(int64(deposit.Till)))
+}
+
+// verify checks whether the transaction was signed by one of the notaries.
+func (n *Notary) verify(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	sig, err := ic.VM.Estack().Pop().Array()[0].TryBytes()
+	if err != nil {
+		panic(fmt.Errorf("failed to get signature bytes: %w", err))
+	}
+	tx := ic.Tx
+	var (
+		signedByNotary, hasEnoughFunds bool
+		needFee                        *big.Int
+	)
+	for _, signer := range tx.Signers {
+		if signer.Account == n.Hash {
+			signedByNotary = true
+			if signer.Scopes != transaction.None {
+				return stackitem.NewBool(false)
+			}
+			if signer.Account.Equals(tx.Sender()) {
+				if len(tx.Signers) == 1 {
```

Comment:
Maybe even `!= 2`, we need one cosigner that pays the bills actually.

Reply:
Fixed. Do the users of this service know that there should be not more than 2 signers?

Reply:
> fallback tx
This transaction has P2P Notary contract as a sender and service request sender as an additional signer.

Already in neo-project/neo#1573 and this is gonna be something to check for in #1547 (#1546 can be little more permissive to simplify things for non-notary nodes).